// @vue-macros/api/dist/ts.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
















var _chunk5TD4KD7Sjs = require('./chunk-5TD4KD7S.js');

















exports.exportsSymbol = _chunk5TD4KD7Sjs.exportsSymbol; exports.getTSFile = _chunk5TD4KD7Sjs.getTSFile; exports.isTSDeclaration = _chunk5TD4KD7Sjs.isTSDeclaration; exports.isTSExports = _chunk5TD4KD7Sjs.isTSExports; exports.mergeTSProperties = _chunk5TD4KD7Sjs.mergeTSProperties; exports.resolveTSEntityName = _chunk5TD4KD7Sjs.resolveTSEntityName; exports.resolveTSExports = _chunk5TD4KD7Sjs.resolveTSExports; exports.resolveTSFileId = _chunk5TD4KD7Sjs.resolveTSFileId; exports.resolveTSFileIdNode = _chunk5TD4KD7Sjs.resolveTSFileIdNode; exports.resolveTSProperties = _chunk5TD4KD7Sjs.resolveTSProperties; exports.resolveTSReferencedType = _chunk5TD4KD7Sjs.resolveTSReferencedType; exports.resolveTSScope = _chunk5TD4KD7Sjs.resolveTSScope; exports.resolveTypeElements = _chunk5TD4KD7Sjs.resolveTypeElements; exports.setResolveTSFileIdImpl = _chunk5TD4KD7Sjs.setResolveTSFileIdImpl; exports.tsFileCache = _chunk5TD4KD7Sjs.tsFileCache; exports.tsFileExportsCache = _chunk5TD4KD7Sjs.tsFileExportsCache;
exports.default = module.exports;
