// @vue-macros/define-slots/dist/chunk-W7M374GL.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkWTXPC3Y4js = require('./chunk-WTXPC3Y4.js');

// src/index.ts
var _unplugin = require('unplugin');






var _common = require('@vue-macros/common');
function resolveOption(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC].concat(
      version === 2 && framework === "webpack" ? _common.REGEX_VUE_SUB : []
    ),
    ...options,
    version
  };
}
var name = "unplugin-vue-define-slots";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        return _chunkWTXPC3Y4js.transformDefineSlots.call(void 0, code, id);
      }
    };
  }
);



exports.src_default = src_default;
