// @vue-macros/better-define/dist/chunk-ONNPRGQT.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkQVCPMPXJjs = require('./chunk-QVCPMPXJ.js');

// src/index.ts
var _unplugin = require('unplugin');






var _common = require('@vue-macros/common');
var _api = require('@vue-macros/api');
function resolveOptions(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC].concat(
      version === 2 && framework === "webpack" ? _common.REGEX_VUE_SUB : []
    ),
    isProduction: process.env.NODE_ENV === "production",
    ...options,
    version
  };
}
var name = "unplugin-vue-better-define";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    const { resolve, handleHotUpdate } = _api.RollupResolve.call(void 0, );
    return {
      name,
      enforce: "pre",
      buildStart() {
        if (framework === "rollup" || framework === "vite") {
          _api.setResolveTSFileIdImpl.call(void 0, resolve(this));
        }
      },
      transformInclude(id) {
        return filter(id);
      },
      async transform(code, id) {
        try {
          return await _chunkQVCPMPXJjs.transformBetterDefine.call(void 0, code, id, options.isProduction);
        } catch (err) {
          this.warn(`${name} ${err}`);
          console.warn(err);
        }
      },
      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction;
        },
        handleHotUpdate
      }
    };
  }
);



exports.src_default = src_default;
