// @vue-macros/setup-component/dist/api.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});












var _chunk4PQK4PHQjs = require('./chunk-4PQK4PHQ.js');













exports.SETUP_COMPONENT_ID_REGEX = _chunk4PQK4PHQjs.SETUP_COMPONENT_ID_REGEX; exports.SETUP_COMPONENT_ID_SUFFIX = _chunk4PQK4PHQjs.SETUP_COMPONENT_ID_SUFFIX; exports.SETUP_COMPONENT_SUB_MODULE = _chunk4PQK4PHQjs.SETUP_COMPONENT_SUB_MODULE; exports.SETUP_COMPONENT_TYPE = _chunk4PQK4PHQjs.SETUP_COMPONENT_TYPE; exports.getMainModule = _chunk4PQK4PHQjs.getMainModule; exports.getScopeDecls = _chunk4PQK4PHQjs.getScopeDecls; exports.hotUpdateSetupComponent = _chunk4PQK4PHQjs.hotUpdateSetupComponent; exports.isSubModule = _chunk4PQK4PHQjs.isSubModule; exports.loadSetupComponent = _chunk4PQK4PHQjs.loadSetupComponent; exports.scanSetupComponent = _chunk4PQK4PHQjs.scanSetupComponent; exports.transformPost = _chunk4PQK4PHQjs.transformPost; exports.transformSetupComponent = _chunk4PQK4PHQjs.transformSetupComponent;
exports.default = module.exports;
