// @vue-macros/reactivity-transform/dist/chunk-ROATVHUW.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});





var _chunk7JCSKXMMjs = require('./chunk-7JCSKXMM.js');

// src/index.ts
var _unplugin = require('unplugin');









var _common = require('@vue-macros/common');
function resolveOption(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_SRC_FILE, _common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC].concat(
      version === 2 && framework === "webpack" ? _common.REGEX_VUE_SUB : []
    ),
    exclude: [_common.REGEX_NODE_MODULES],
    ...options,
    version
  };
}
var name = "unplugin-reactivity-transform";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (id === _chunk7JCSKXMMjs.helperId)
          return id;
      },
      loadInclude(id) {
        return id === _chunk7JCSKXMMjs.helperId;
      },
      load(_id) {
        const id = _common.normalizePath.call(void 0, _id);
        if (id === _chunk7JCSKXMMjs.helperId)
          return _chunk7JCSKXMMjs.code_default;
      },
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        if (_common.REGEX_VUE_SFC.test(id) || _common.REGEX_SETUP_SFC.test(id) || framework === "webpack" && _common.REGEX_VUE_SUB.test(id)) {
          return _chunk7JCSKXMMjs.transformVueSFC.call(void 0, code, id);
        } else if (_chunk7JCSKXMMjs.shouldTransform.call(void 0, code)) {
          return _chunk7JCSKXMMjs.transform.call(void 0, code, {
            filename: id,
            sourceMap: true
          });
        }
      }
    };
  }
);



exports.src_default = src_default;
