// @vue-macros/define-models/dist/chunk-3XO4EKGG.mjs
import {
  emitHelperId,
  emit_helper_default,
  helperPrefix,
  transformDefineModels,
  useVmodelHelperId,
  use_vmodel_default
} from "./chunk-6LNWKFVH.mjs";

// src/index.ts
import { createUnplugin } from "unplugin";
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
  normalizePath
} from "@vue-macros/common";
function resolveOption(options, framework) {
  const version = options.version || detectVueVersion();
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === "webpack" ? REGEX_VUE_SUB : []
    ),
    unified: true,
    ...options,
    version
  };
}
var name = "unplugin-vue-define-models";
var src_default = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions, framework);
    const filter = createFilter(options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (id.startsWith(helperPrefix))
          return id;
      },
      loadInclude(id) {
        return id.startsWith(helperPrefix);
      },
      load(_id) {
        const id = normalizePath(_id);
        if (id === emitHelperId)
          return emit_helper_default;
        else if (id === useVmodelHelperId)
          return use_vmodel_default;
      },
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        return transformDefineModels(code, id, options.version, options.unified);
      }
    };
  }
);

export {
  src_default
};
