// @vue-macros/define-models/dist/chunk-2Z4EHQYL.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});






var _chunkTHNFZ32Ejs = require('./chunk-THNFZ32E.js');

// src/index.ts
var _unplugin = require('unplugin');







var _common = require('@vue-macros/common');
function resolveOption(options, framework) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC].concat(
      version === 2 && framework === "webpack" ? _common.REGEX_VUE_SUB : []
    ),
    unified: true,
    ...options,
    version
  };
}
var name = "unplugin-vue-define-models";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions, framework);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (id.startsWith(_chunkTHNFZ32Ejs.helperPrefix))
          return id;
      },
      loadInclude(id) {
        return id.startsWith(_chunkTHNFZ32Ejs.helperPrefix);
      },
      load(_id) {
        const id = _common.normalizePath.call(void 0, _id);
        if (id === _chunkTHNFZ32Ejs.emitHelperId)
          return _chunkTHNFZ32Ejs.emit_helper_default;
        else if (id === _chunkTHNFZ32Ejs.useVmodelHelperId)
          return _chunkTHNFZ32Ejs.use_vmodel_default;
      },
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        return _chunkTHNFZ32Ejs.transformDefineModels.call(void 0, code, id, options.version, options.unified);
      }
    };
  }
);



exports.src_default = src_default;
