// @vue-macros/common/dist/ast.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
















var _chunkMPXRPYOHjs = require('./chunk-MPXRPYOH.js');
require('./chunk-TACQCSM2.js');
require('./chunk-5WEO6FF7.js');

















exports.HELPER_PREFIX = _chunkMPXRPYOHjs.HELPER_PREFIX; exports.TS_NODE_TYPES = _chunkMPXRPYOHjs.TS_NODE_TYPES; exports.babelParse = _chunkMPXRPYOHjs.babelParse; exports.checkInvalidScopeReference = _chunkMPXRPYOHjs.checkInvalidScopeReference; exports.importHelperFn = _chunkMPXRPYOHjs.importHelperFn; exports.isCallOf = _chunkMPXRPYOHjs.isCallOf; exports.isFunctionType = _chunkMPXRPYOHjs.isFunctionType; exports.isLiteralType = _chunkMPXRPYOHjs.isLiteralType; exports.isStaticExpression = _chunkMPXRPYOHjs.isStaticExpression; exports.isStaticObjectKey = _chunkMPXRPYOHjs.isStaticObjectKey; exports.resolveLiteral = _chunkMPXRPYOHjs.resolveLiteral; exports.resolveObjectExpression = _chunkMPXRPYOHjs.resolveObjectExpression; exports.resolveObjectKey = _chunkMPXRPYOHjs.resolveObjectKey; exports.resolveTemplateLiteral = _chunkMPXRPYOHjs.resolveTemplateLiteral; exports.unwrapTSNode = _chunkMPXRPYOHjs.unwrapTSNode; exports.walkAST = _chunkMPXRPYOHjs.walkAST;
exports.default = module.exports;
