// @vue-macros/common/dist/constants.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});






















var _chunk5WEO6FF7js = require('./chunk-5WEO6FF7.js');























exports.DEFINE_EMIT = _chunk5WEO6FF7js.DEFINE_EMIT; exports.DEFINE_EMITS = _chunk5WEO6FF7js.DEFINE_EMITS; exports.DEFINE_MODELS = _chunk5WEO6FF7js.DEFINE_MODELS; exports.DEFINE_MODELS_DOLLAR = _chunk5WEO6FF7js.DEFINE_MODELS_DOLLAR; exports.DEFINE_OPTIONS = _chunk5WEO6FF7js.DEFINE_OPTIONS; exports.DEFINE_PROP = _chunk5WEO6FF7js.DEFINE_PROP; exports.DEFINE_PROPS = _chunk5WEO6FF7js.DEFINE_PROPS; exports.DEFINE_PROPS_DOLLAR = _chunk5WEO6FF7js.DEFINE_PROPS_DOLLAR; exports.DEFINE_PROPS_REFS = _chunk5WEO6FF7js.DEFINE_PROPS_REFS; exports.DEFINE_RENDER = _chunk5WEO6FF7js.DEFINE_RENDER; exports.DEFINE_SETUP_COMPONENT = _chunk5WEO6FF7js.DEFINE_SETUP_COMPONENT; exports.DEFINE_SLOTS = _chunk5WEO6FF7js.DEFINE_SLOTS; exports.REGEX_DTS = _chunk5WEO6FF7js.REGEX_DTS; exports.REGEX_LANG_JSX = _chunk5WEO6FF7js.REGEX_LANG_JSX; exports.REGEX_LANG_TS = _chunk5WEO6FF7js.REGEX_LANG_TS; exports.REGEX_NODE_MODULES = _chunk5WEO6FF7js.REGEX_NODE_MODULES; exports.REGEX_SETUP_SFC = _chunk5WEO6FF7js.REGEX_SETUP_SFC; exports.REGEX_SRC_FILE = _chunk5WEO6FF7js.REGEX_SRC_FILE; exports.REGEX_VUE_SFC = _chunk5WEO6FF7js.REGEX_VUE_SFC; exports.REGEX_VUE_SUB = _chunk5WEO6FF7js.REGEX_VUE_SUB; exports.REPO_ISSUE_URL = _chunk5WEO6FF7js.REPO_ISSUE_URL; exports.WITH_DEFAULTS = _chunk5WEO6FF7js.WITH_DEFAULTS;
exports.default = module.exports;
