// @vue-macros/devtools/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }// src/index.ts
var _path = require('path');
var _sirv = require('sirv'); var _sirv2 = _interopRequireDefault(_sirv);
var DEV_SERVER_PATH = "/__vue-macros";
var Devtools = ({ nuxtContext } = {}) => {
  return {
    name: "vue-macros-devtools",
    async configureServer(server) {
      if ((nuxtContext == null ? void 0 : nuxtContext.isClient) === false)
        return;
      if (false) {
        const { createServer } = await null;
        const subServer = await createServer({
          root: _path.resolve.call(void 0, __dirname, "../src/client"),
          server: {
            hmr: {
              port: await getPort()
            },
            middlewareMode: true
          }
        });
        server.middlewares.use(DEV_SERVER_PATH, subServer.middlewares);
      } else {
        server.middlewares.use(
          DEV_SERVER_PATH,
          _sirv2.default.call(void 0, _path.resolve.call(void 0, __dirname, "client"), {
            single: true,
            dev: true
          })
        );
      }
    }
  };
};


exports.Devtools = Devtools;
