// @vue-macros/devtools/dist/index.mjs
// ../../node_modules/.pnpm/tsup@6.6.3_ujzcdj5eu57huy77ckcec3nqfu/node_modules/tsup/assets/esm_shims.js
import { fileURLToPath } from "url";
import path from "path";
var getFilename = () => fileURLToPath(import.meta.url);
var getDirname = () => path.dirname(getFilename());
var __dirname = /* @__PURE__ */ getDirname();

// src/index.ts
import { resolve } from "path";
import sirv from "sirv";
var DEV_SERVER_PATH = "/__vue-macros";
var Devtools = ({ nuxtContext } = {}) => {
  return {
    name: "vue-macros-devtools",
    async configureServer(server) {
      if ((nuxtContext == null ? void 0 : nuxtContext.isClient) === false)
        return;
      if (false) {
        const { createServer } = await null;
        const subServer = await createServer({
          root: resolve(__dirname, "../src/client"),
          server: {
            hmr: {
              port: await getPort()
            },
            middlewareMode: true
          }
        });
        server.middlewares.use(DEV_SERVER_PATH, subServer.middlewares);
      } else {
        server.middlewares.use(
          DEV_SERVER_PATH,
          sirv(resolve(__dirname, "client"), {
            single: true,
            dev: true
          })
        );
      }
    }
  };
};
export {
  Devtools
};
